### PR TITLE
[Blog Post] Instrumenting Phoenix with Telemetry Part 2

### PR DESCRIPTION
**Pull request by SophieDeBenedetto**
Part 2 of the series outlined in this PR https://github.com/elixirschool/elixirschool/pull/2249

### CODE REVIEW

File: _posts/2020-04-29-instrumenting_phoenix_with_telemetry_part_two.md [hunk 1]
```diff
@@ -0,0 +1,450 @@
...
+end
+```
+
+## What's Wrong With This?
+
+Telemetry made it easy for us to emit an event and operate on it, but our current usage of the Telemetry library leaves a lot to be desired.
+
+One drawback of our current approach is that it leaves us on the hook for Telemetry event handling and metrics reporting. We had to define our own custom event handling module, manually attach that module to the given Telemetry event and define the handler's callback function.
+
+In order for that callback function report metrics to StatsD for a given event, we had to create our own custom module that uses the `Statix` library _and_ write code that formats the metric to send to StatsD for a given Telemetry event. The mental overhead of translating Telemetry event data into the appropriate StatsD metric is costly, and that effort will have to be undertaken for every new Telemetry event we execute and handle.
```
Comment (jordelver) [inline _posts/2020-04-29-instrumenting_phoenix_with_telemetry_part_two.md:62]:
```suggestion
In order for that callback function to report metrics to StatsD for a given event, we had to create our own custom module that uses the `Statix` library _and_ write code that formats the metric to send to StatsD for a given Telemetry event. The mental overhead of translating Telemetry event data into the appropriate StatsD metric is costly, and that effort will have to be undertaken for every new Telemetry event we execute and handle.
```

File: _posts/2020-04-29-instrumenting_phoenix_with_telemetry_part_two.md [hunk 2]
```diff
@@ -0,0 +1,450 @@
...
+
+Later, when a Telemetry event is executed via a call to `:telemetry.execute/3`, Telemetry looks up the event handler, `TelemetryMetricsStatsd.EventHandler`, for the given event in the ETS table and invokes it. The event handler module will format the event, metadata and any associated tags as the appropriate StatsD metric and send the resulting metric to StatsD over UDP.
+
+Most of this happens under the hood. We are only on the hook for defining a `Telemetry.Metrics` module and listing the Telemetry events we want to handle as which type of metric. That's it!
+
+## Getting Started
+
+You can follow along with this tutorial by cloning down the repo [here](https://github.com/elixirschool/telemetry-code-along/tree/part-2-start).
+* Checkig out the starting state of our code on the branch [part-2-start](https://github.com/elixirschool/telemetry-code-along/tree/part-2-start)
```
Comment (jordelver) [inline _posts/2020-04-29-instrumenting_phoenix_with_telemetry_part_two.md:99]:
```suggestion
* Checking out the starting state of our code on the branch [part-2-start](https://github.com/elixirschool/telemetry-code-along/tree/part-2-start)
```

File: _posts/2020-04-29-instrumenting_phoenix_with_telemetry_part_two.md [hunk 3]
```diff
@@ -0,0 +1,450 @@
...
+        tags: [:request_path]
+      )
+    ]
+  end
+end
+```
+
+This kicks off the following process:
+
+* When the `TelemetryMetricsStatsd` starts, it stores events in ETS along with their handler _and_ a config map including this list of metric structs.
```
Comment (jordelver) [inline _posts/2020-04-29-instrumenting_phoenix_with_telemetry_part_two.md:275]:
```suggestion
* When the `TelemetryMetricsStatsd` starts, it stores events in ETS along with their handler _and_ a config map including this list of metric structs
```

File: _posts/2020-04-29-instrumenting_phoenix_with_telemetry_part_two.md [hunk 4]
```diff
@@ -0,0 +1,450 @@
...
+```elixir
+# lib/quantum_web/controllers/user_controller.ex
+def new(conn, _params)
+  :telemetry.execute([:phoenix, :request], %{duration: System.monotonic_time() - start}, conn)
+end
+```
+
+The `TelemetryMetricsStatsD` will format a counter and summary metric tagged with the value of the `:request_path` key found in the `conn`.
+
+
```
Comment (jordelver) [inline _posts/2020-04-29-instrumenting_phoenix_with_telemetry_part_two.md:293]:
```suggestion
```

File: _posts/2020-04-29-instrumenting_phoenix_with_telemetry_part_two.md [hunk 5]
```diff
@@ -0,0 +1,450 @@
...
+}
+```
+
+## Under The Hood
+
+The `Quantum.Telemetry` module is, believe it or not, the _only_ code we have to write in order to send these metrics to StatsD for the `"phoenix.router_dispatch.stop"` event. The Telemetry libraries take care of everything else for us under the hood.
+
+Let's take a closer look at how it all works.
+
+1. The `Telemetry.Metrics` supervisor that we defined in `Quantum.Telemetry` defines a list of metrics that we want to emit to StatsD for a given Telemetry event.
```
Comment (jordelver) [inline _posts/2020-04-29-instrumenting_phoenix_with_telemetry_part_two.md:331]:
```suggestion
1. The `Telemetry.Metrics` supervisor that we defined in `Quantum.Telemetry` defines a list of metrics that we want to emit to StatsD for a given Telemetry event
```

File: _posts/2020-04-29-instrumenting_phoenix_with_telemetry_part_two.md [hunk 6]
```diff
@@ -0,0 +1,450 @@
...
+2. When the supervisor starts, it starts the `TelemetryMetricsStatsd` GenServer and gives it this list
+3. When the `TelemetryMetricsStatsd` GenServer starts, it calls `:telemetry.attach/4` for each listed event, storing it in an ETS table along with the handler callback and a config map that includes the metrics definitions. The handler callback it gives to `:telemetry.attach/4` is its own `TelemetryMetricsStatsd.EventHandler.handle_event/4` function.
+4. Later, when a Telemetry event is executed via a call to `:telemetry.execute/3`, Telemetry looks up the handler callback and config (including metrics definitions) for the given event in ETS.
```
Comment (jordelver) [inline _posts/2020-04-29-instrumenting_phoenix_with_telemetry_part_two.md:334]:
```suggestion
4. Later, when a Telemetry event is executed via a call to `:telemetry.execute/3`, Telemetry looks up the handler callback and config (including metrics definitions) for the given event in ETS
```

File: _posts/2020-04-29-instrumenting_phoenix_with_telemetry_part_two.md [hunk 7]
```diff
@@ -0,0 +1,450 @@
...
+5. The `:telemetry.execute/3` function then calls this handler callback, `TelemetryMetricsStatsd.EventHandler.handle_event/4`, with the event name, event measurement map, event metadata and metrics config.
```
Comment (jordelver) [inline _posts/2020-04-29-instrumenting_phoenix_with_telemetry_part_two.md:335]:
```suggestion
5. The `:telemetry.execute/3` function then calls this handler callback, `TelemetryMetricsStatsd.EventHandler.handle_event/4`, with the event name, event measurement map, event metadata and metrics config
```

**Comment by SophieDeBenedetto** [comment]
@jordelver thanks for your fixes--merged them in 😄  I'm hoping to publish this tomorrow to stick to a one week cadence of releasing each addition in this series. So, if it otherwise looks good to you and you want to go ahead and hit 'approve', that would be great! 🚀

**Comment by jordelver** [comment]
@SophieDeBenedetto thanks for writing these. I'm learning a lot :)

**Comment by jordelver** [comment]
Looks like you need approval from Developers or English teams before it can be merged?

**Pull request merged**